Mathlib/Order/RelSeries.lean

```diff
@@ -0,0 +1,261 @@
+abbrev toList (x : RelSeries r) : List α := List.ofFn x
+
+lemma toList_chain' (x : RelSeries r) : x.toList.Chain' r := by
+  rw [List.chain'_iff_get]
+  intros i h
+  have h' : i < x.length := by simpa [List.length_ofFn] using h
+  convert x.step ⟨i, h'⟩ <;>
+  · rw [List.get_ofFn]
+    congr 1
+
+lemma toList_ne_empty (x : RelSeries r) : x.toList ≠ ∅ := fun m =>
+  List.eq_nil_iff_forall_not_mem.mp m (x 0) <| (List.mem_ofFn _ _).mpr ⟨_, rfl⟩
+
+/-- Every nonempty list satisfying the chain condition gives a relation series-/
+@[simps]
+def fromListChain' (x : List α) (x_ne_empty : x ≠ ∅) (hx : x.Chain' r) : RelSeries r where
+  length := x.length.pred
+  toFun := x.get ∘ Fin.cast (Nat.succ_pred_eq_of_pos <| List.length_pos.mpr x_ne_empty)
+  step := fun i => List.chain'_iff_get.mp hx i i.2
+
+/-- Relation series of `r` and nonempty list of `α` satisfying `r`-chain condition bijectively
+corresponds to each other.-/
+protected def Equiv : RelSeries r ≃ {x : List α | x ≠ ∅ ∧ x.Chain' r} where
+  toFun := fun x => ⟨_, x.toList_ne_empty, x.toList_chain'⟩
+  invFun := fun x => fromListChain' _ x.2.1 x.2.2
+  left_inv := fun x => ext (by dsimp; rw [List.length_ofFn, Nat.pred_succ]) <| by
+    ext f
+    simp only [List.empty_eq, ne_eq, Set.coe_setOf, Set.mem_setOf_eq, fromListChain'_toFun,
+      Function.comp_apply, List.get_ofFn, fromListChain'_length]
+    rfl
+  right_inv := by
+    intro x
+    refine Subtype.ext (List.ext_get ?_ <| fun n hn1 hn2 => ?_)
+    · dsimp
+      rw [List.length_ofFn, fromListChain'_length, ←Nat.succ_eq_add_one, Nat.succ_pred_eq_of_pos]
+      rw [List.length_pos]
+      exact x.2.1
+    · rw [List.get_ofFn, fromListChain'_toFun, Function.comp_apply]
+      congr
+
+-- TODO : build a similar bijection between `RelSeries α` and `Quiver.Path`
+
+end RelSeries
+
+namespace Rel
+
+/-- a relation `r` is said to be finite dimensional iff there is a relation series of `r` with the
+  maximum length. -/
+class FiniteDimensional where
+  /-- the longest relation series-/
+  longestRelSeries : RelSeries r
+  /-- the longest relation series is longer than any other series -/
+  is_longest : ∀ (y : RelSeries r), y.length ≤ longestRelSeries.length
+
+/-- the longest relational series when a relation is finite dimensional -/
+def longestRelSeries [r.FiniteDimensional] : RelSeries r := FiniteDimensional.longestRelSeries
+
+lemma longestRelSeries_is_longest [r.FiniteDimensional] (x : RelSeries r) :
+  x.length ≤ r.longestRelSeries.length := FiniteDimensional.is_longest _
+
+/-- a relation `r` is said to be infinite dimensional iff there exists relation series of arbitrary
+  length. -/
+class InfiniteDimensional where
+  /-- there should a relation series with length at least `n` for all natural number `n` -/
+  longRelSeries : ℕ → RelSeries r
+  /-- there should a relation series with length at least `n` for all natural number `n` -/
+  longRelSeries_length : ∀ (n : ℕ), n ≤ (longRelSeries n).length
```

Comment:
Similarly, I would define this class to consists of a single field `exists_length_ge : ∀ n : ℕ, ∃ s : RelSeries r, n ≤ s.length`.

Reply:
changed to this version